### PR TITLE
Add RuntimeError in _distn_infrastructure.py in fit() method 

### LINKED ISSUE #10300: scipy.stats.rv_continuous.fit throws empty RuntimeError when input has NaNs
When using the `scipy.stats.rv_continuous.fit` function, a `RuntimeError` (without a message) is thrown if the array to which the distribution is being fit contains `NaN`s. It would be nice to have a warning (or error) for NaNs.

### Reproducing code example:

```python
import numpy as np
import scipy.stats as stats

x = np.array([1.6483, 1.8223, 2.7169, 2.4667, 1.1791, 3.5433, 2.6561, 1.6631,
       2.412 , 6.3796, 1.7192, 1.438 , 4.5905, 2.832 , 1.86  , 2.1236,
       2.5366, 2.442 , 1.7216, 1.6853, 2.4713, 1.7739, 3.3199, 3.711 ,
       1.821 , 1.6071, 3.9321, 3.4891, 3.0009, 2.0428, 2.6311, 2.218 ,
       2.3838, 2.6662, 2.3853, 2.3079, 3.4989, 4.8993, 8.8305, 3.6952,
       1.9436, 5.7345, 3.1823, 2.8493, 2.2327, 3.2036, 2.3663, 3.0886,
       2.4526, 2.0095, 2.6788, 4.3908, 2.6042, 1.1514, 3.6923, 2.5201,
       2.8638, 1.7267, 6.2628, 2.3748, 2.3494, 4.6349, 2.8303, 2.1384,
       5.0081, 2.3307, 3.543 , 1.8633, 2.3803, 2.0992, 1.8173, 1.6052,
       2.4661, 3.3964, 2.4746, 2.1748, 1.673 , 3.1572, 3.7291, 2.2546,
       1.5754, 2.2669, 1.6383, 1.7696, 1.9079, 1.0955, 1.5637, 2.66  ,
       1.746 , 1.5319, 2.0958,    np.nan])
K = stats.exponnorm.fit(x)
```

### Error message:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-86-fb3d43a0acaa> in <module>
     13        1.5754, 2.2669, 1.6383, 1.7696, 1.9079, 1.0955, 1.5637, 2.66  ,
     14        1.746 , 1.5319, 2.0958,    np.nan])
---> 15 K = stats.exponnorm.fit(x)

/anaconda3/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py in fit(self, data, *args, **kwds)
   2195                                          'scale' in kwds):
   2196             # get distribution specific starting locations
-> 2197             start = self._fitstart(data)
   2198             args += start[Narg:-2]
   2199         loc = kwds.pop('loc', start[-2])

/anaconda3/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py in _fitstart(self, data, args)
   2034         if args is None:
   2035             args = (1.0,)*self.numargs
-> 2036         loc, scale = self._fit_loc_scale_support(data, *args)
   2037         return args + (loc, scale)
   2038 

/anaconda3/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py in _fit_loc_scale_support(self, data, *args)
   2287             return (data_b - b) + margin, 1
   2288         else:
-> 2289             raise RuntimeError
   2290 
   2291     def fit_loc_scale(self, data, *args):

RuntimeError: 
```

### Scipy/Numpy/Python version information:
```
1.2.1 1.15.4 sys.version_info(major=3, minor=7, micro=1, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by rlucas7**
This PR is _one_ proposal to fix #10300 from the discussion w/ @chrisb83  @WarrenWeckesser  and  @kmundnic (who opened the issue). From the discussion on the issue thread it is not clear to me if we want to implement a `nan_policy` for fit or simply throw an error. Given the inconsistency in the current behavior any fix will change the behavior of the `fit()` method. I'd rather have the fix w/minimal disruption to the user base, not sure if this approach or `nan_policy` has less disruption. This implementation was quicker to put together than a full `nan_policy`. However, if we want a full `nan_policy` we should close this PR w/o merging and I'll work towards implementing a `nan_policy`. The 3 commits here are minimal, and replicate the issue in #10300 and resolve the issue by throwing a `RuntimeError` with a descriptive message of the issue.

### CODE REVIEW

File: scipy/stats/_distn_infrastructure.py [hunk 1]
```diff
@@ -2256,6 +2258,9 @@ def fit(self, data, *args, **kwds):
         if Narg > self.numargs:
             raise TypeError("Too many input arguments.")
 
+        if np.logical_not(np.isfinite(data)).any():
+            raise RuntimeError("The data contains np.nan values.")
```
Comment (chrisb83) [inline scipy/stats/_distn_infrastructure.py]: "contains non-finite values" is better

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -1412,6 +1412,11 @@ def get_moms(lam, sig, mu):
         sts = stats.exponnorm.stats(K, loc=mu, scale=sig, moments='mvsk')
         assert_almost_equal(sts, get_moms(lam, sig, mu))
 
+    def test_npnan_raises_error(self):
+        with pytest.raises(RuntimeError) as excinfo:
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: you can use `assert_raises`, that is aligned to other tests

Maybe add another test for `np.inf`?

**Review by chrisb83 — COMMENTED** [review]
only two minor comments

**Comment by chrisb83** [comment]
> it is not clear to me if we want to implement a nan_policy for fit or simply throw an error.

My impression was that throwing an error was the prefered option. (`nan_policy` has quite a few issues)

**Comment by WarrenWeckesser** [comment]
It will probably be a week or so before I can give a real review of this.  For now I'll just note that several of the distributions (e.g. `norm`, `beta` and a few others) override the `fit` method, so they will have to be updated individually.

**Comment by rlucas7** [comment]
>My impression was that throwing an error was the prefered option. (nan_policy has quite a few issues)

@chrisb83 after reading again it is clear to me the consensus on throwing an error. Not sure why I hadn't caught that before.

>It will probably be a week or so before I can give a real review of this. For now I'll just note that several of the distributions (e.g. norm, beta and a few others) override the fit method, so they will have to be updated individually.

I know you're busy @WarrenWeckesser and your comments are always helpful. I added some lines into the overridden `.fit()` methods and added full distributional unit test coverage.

File: scipy/stats/tests/test_distributions.py (hunk 2 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: added a test for `np.inf`

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_distn_infrastructure.py]: changes the message in error on recent commits (in all occurrences)

**Comment by rlucas7** [comment]
@chrisb83 could we also add a 1.4.0 release tag on this so we at least revisit before the next release? (hopefully gets in sooner).

**Comment by tylerjereddy** [comment]
> could we also add a 1.4.0 release tag on this so we at least revisit before the next release? 

Yes, done

**Comment by rlucas7** [comment]
> > could we also add a 1.4.0 release tag on this so we at least revisit before the next release?
> 
> Yes, done

Thanks @tylerjereddy !

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -264,6 +264,9 @@ def fit(self, data, **kwds):
 
         data = np.asarray(data)
 
+        if np.logical_not(np.isfinite(data)).any():
```
Comment (pv) [inline scipy/stats/_continuous_distns.py]: Drive-by-comment: `not np.isfinite(data).all()` is likely faster since `logical_not` generates an intermediate array.

Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: > Drive-by-comment: `not np.isfinite(data).all()` is likely faster since `logical_not` generates an intermediate array.

I bet this matters more on larger arrays. I was surprised to learn even on the test case there is a speedup: 
```
import scipy.stats as stats
import numpy as np
from datetime import datetime
startTime = datetime.now()
for i in  range(0, 10000):
    x = np.array([1.6483, 2.7169, 2.4667, 1.1791, 3.5433, np.nan])          
    np.testing.assert_raises(RuntimeError, stats.norm.fit, x)

print(datetime.now() - startTime)
```
gives runtime of: 
# 0:00:00.156653 (old) 
# 0:00:00.138408 (proposed)

for a larger range 10000 -> 1000000 the improvements are: 
0:00:12.887804 (old) 
0:00:12.201071 (new)   
 
I'll change all the implementations to that proposed by @pv  Thanks!

**Comment by rlucas7** [comment]
I just rebased the branch off master, commit `2b41710`, the branch was pretty stale (~662 commits behind master). There was only 1 conflict in the `TestFitMethod` looks like some things were either removed or factored out to a different file. After the conflict resolutions the `stats` tests all pass on my local.

**Comment by rlucas7** [comment]
> I just rebased the branch off master, commit `2b41710`, the branch was pretty stale (~662 >commits behind master). There was only 1 conflict in the `TestFitMethod` looks like some things >were either removed or factored out to a different file. After the conflict resolutions the `stats` >tests all pass on my local. 
>Would be great to get this into the 1.4.0 release.

I also added a backport candidate label. I'm still unclear on what items we usually backport. This one seems like a good candidate b/c it makes inconsistent behavior in the `fit()` methods consistent. If the committer thinks it isn't a backport candidate that's fine too but figured I'd label so we consider it for the backport.

**Comment by tylerjereddy** [comment]
>  I'm still unclear on what items we usually backport

Usually just bug fixes--is this bad enough to be considered a bug?

**Comment by rlucas7** [comment]
> > I'm still unclear on what items we usually backport
> 
> Usually just bug fixes--is this bad enough to be considered a bug?

I'm not sure, the behavior on some fit methods if `nan` or `inf` is part of fitting data is inconsistent across  distributions. 
Sounds like this isn't a bug, it isn't too serious, whatever behavior a user is seeing they've likely been seeing for awhile.
I'll remove the backport label, thanks for clarifying @tylerjereddy

**Comment by tylerjereddy** [comment]
@WarrenWeckesser Any concerns here? It looks like you were planning to review but didn't get around to it. Beyond that, it seems reviewer comments were addressed.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
+1 for the consistent handling of non-finite values.  I'm OK with raising an exception.  This leaves open the possibility of adding the `nan_policy` argument sometime in the future.  One thing to be aware of: the change here raises an exception on `nan` or `inf`.  `nan_policy` has been specifically about `nan`, not `inf`.  I think expanding the `nan_policy` to include handling of `inf` in a consistent and backwards-compatible way is something we can do.  I'm just noting that once we commit here to raising on `inf`, a future addition of a `nan_policy` argument must provide an option for raising errors on `inf` (along with `nan`).  That is, we'll need an API for `nan_policy` that has an option for erroring on just `nan` or for erroring on any non-finite values, and the `nan_policy` for the `fit` method would need to use the option "raise on `nan` or `inf`" as its default.

I noted one issue in the test code to be addressed.

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -1842,6 +1888,22 @@ class TestFitMethod(object):
...
 
+    # skip these b/c depracated, or only loc and scale arguments
+    fitSkipNonFinite = ['frechet_l', 'frechet_r', 'expon', 'norm', 'uniform', ]
+
+    @pytest.mark.parametrize('dist,args', distcont)
+    def test_fit_w_non_finite_data_values(self, dist, args):
+        """gh-10300"""
+        if dist in self.fitSkipNonFinite:
+            pytest.skip("%s fit known to fail or depracated" % dist)
+        with np.errstate(all='ignore'), suppress_warnings() as sup:
```
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: @WarrenWeckesser thanks! You are right, the context manager is not needed. This was a copy paste error IIRC. Doesn't add anything, removed. If the tests all pass I can do a rebase too, this branch is now fairly stale.

**Comment by rlucas7** [comment]
> +1 for the consistent handling of non-finite values. I'm OK with raising an exception. This leaves open the possibility of adding the `nan_policy` argument sometime in the future. One thing to be aware of: the change here raises an exception on `nan` or `inf`. `nan_policy` has been specifically about `nan`, not `inf`. I think expanding the `nan_policy` to include handling of `inf` in a consistent and backwards-compatible way is something we can do. I'm just noting that once we commit here to raising on `inf`, a future addition of a `nan_policy` argument must provide an option for raising errors on `inf` (along with `nan`). That is, we'll need an API for `nan_policy` that has an option for erroring on just `nan` or for erroring on any non-finite values, and the `nan_policy` for the `fit` method would need to use the option "raise on `nan` or `inf`" as its default.

I agree that we likely want to have something to handle `nan` and something to handle `inf`. 
I'm not sure that `nan_policy` is the best name for it though :). 
Maybe we could factor out an `inf_policy` and a `nan_policy`?
Seems a bit weird to label `nan_policy` if it is really `nan` ***and*** `inf` that are handled by the method. 
Maybe I'm splitting hairs here but it could be a `non_finite` policy which delegates to `nan_policy` and `inf_policy`.

**Comment by rlucas7** [comment]
Surprisingly no conflicts on rebase and perhaps my "largest" rebase yet:
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ git rev-list e90bafdfe1fa7266f3e3befc7f82368ce38bdf9f   2b417108907ff3254fe1 | wc -l 
   22226
```

**Comment by rlucas7** [comment]
@tylerjereddy @WarrenWeckesser I've removed the context manager as requested. Please let me know if you'd like anything else changed. 

We can take the `nan_policy` naming discussion to the dev-list.

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -2020,6 +2066,20 @@ class TestFitMethod(object):
     def setup_method(self):
         np.random.seed(1234)
 
+    # skip these b/c depracated, or only loc and scale arguments
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py]: Spelling: deprecated

File: scipy/stats/tests/test_distributions.py [hunk 6]
```diff
@@ -2020,6 +2066,20 @@ class TestFitMethod(object):
...
+    fitSkipNonFinite = ['frechet_l', 'frechet_r', 'expon', 'norm', 'uniform', ]
+
+    @pytest.mark.parametrize('dist,args', distcont)
+    def test_fit_w_non_finite_data_values(self, dist, args):
+        """gh-10300"""
+        if dist in self.fitSkipNonFinite:
+            pytest.skip("%s fit known to fail or depracated" % dist)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py]: Spelling: deprecated

File: scipy/stats/tests/test_distributions.py (hunk 5 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: done

File: scipy/stats/tests/test_distributions.py (hunk 6 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: done.

**Review by WarrenWeckesser — APPROVED** [review]
Thanks Lucas, looks good.

**Comment by rlucas7** [comment]
Thanks Warren. I added a brief note under backwards incompatible changes in the release notes wiki.
